src/blockdata/transaction.rs

```diff
@@ -888,15 +917,44 @@ impl crate::consensus::encode::Encodable for SubField {
                 Ok(len + nonce.consensus_encode(w)?)
             }
             SubField::MergeMining(ref depth, ref merkle_root) => {
-                len += 0x3u8.consensus_encode(w)?;
-                match depth {
-                    Some(dep) => {
-                        len += VarInt(33).consensus_encode(w)?;
-                        len += dep.consensus_encode(w)?;
-                    }
-                    None => len += VarInt(32).consensus_encode(w)?,
+                // As per monero code `release-v0.18`, `bool add_mm_merkle_root_to_tx_extra(...)`
+                #[cfg(not(feature = "merge_mining_field_upgrade_v0_18"))]
+                {
+                    // Note: If `depth_to_encode >= VarInt(32)` it will not be acceptable in monero, but here the code will
+                    // not error; user applications have to ensure that such values are not passed in.
+                    let depth_to_encode = match depth {
+                        Some(dep) => dep.clone(),
+                        None => VarInt(0),
+                    };
+                    // Tag
+                    len += 0x3u8.consensus_encode(w)?;
+                    // Data size
+                    len += 33u8.consensus_encode(w)?;
+                    // Depth
+                    len += depth_to_encode.0.to_le_bytes()[0].consensus_encode(w)?;
+                    // Hash
+                    Ok(len + merkle_root.consensus_encode(w)?)
+                }
+                // As per monero code after `release-v0.18`, `bool add_mm_merkle_root_to_tx_extra(...)`
+                #[cfg(feature = "merge_mining_field_upgrade_v0_18")]
+                {
+                    let depth_to_encode = match depth {
+                        Some(dep) => dep.clone(),
+                        None => VarInt(0),
+                    };
+                    // Tag (u8)
+                    len += 0x3u8.consensus_encode(w)?;
+                    // Data size (u8)
+                    let mut writer_temp = vec![];
+                    let depth_var_int_size = depth_to_encode.consensus_encode(&mut writer_temp)?;
+                    let data_length = 1 /*tag size*/ + 1 /*data size*/ + 32 /*hash*/ + depth_var_int_size.to_le_bytes()[0];
+                    len += data_length.consensus_encode(w)?;
+                    // Depth (VarInt)
+                    len += depth_to_encode.consensus_encode(w)?;
+                    // Hash ([u8; 32])
+                    len += merkle_root.consensus_encode(w)?;
+                    Ok(len)
```

Comment:
`data_length` here should just be 32 + size of varint. We could probably simplify this whole section to:

```suggestion
                    len += 0x3u8.consensus_encode(w)?;
                    let mut writer_temp = vec![];

                    depth_to_encode.consensus_encode(&mut writer_temp)?;
                    merkle_root.consensus_encode(&mut writer_temp)?;

                    len += writer_temp.consensus_encode(w)?;
                    Ok(len)
```

Reply:
Yes, I misread the `data_length` encoding